### DISCUSSION TITLE
Auto memory woes – Python exceptions and DNS lookup errors?

### DISCUSSION THREAD

**Discussion by joestump** [open-webui/open-webui#7220] [Issues]
👋🏼 Hey all, trying hard to wrap my brain around functions. Thought I'd start simple with the [`auto_memory`](https://openwebui.com/f/caplescrest/auto_memory) function, but am running into some strange issues. 

I am regularly seeing things like:

```
HTTPConnectionPool(host='host.docker.internal', port=11434): Max retries exceeded with url: /api/embed (Caused by NameResolutionError("<urllib3.connection.HTTPConnection object at 0x7137593ad850>: Failed to resolve 'host.docker.internal' ([Errno -2] Name or service not known)"))
```

And 

```
INFO:     10.11.12.4:0 - "POST /api/v1/memories/query HTTP/1.1" 500 Internal Server Error
ERROR:    Exception in ASGI application
  + Exception Group Traceback (most recent call last):
  |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/base.py", line 192, in __call__
  |     await response(scope, wrapped_receive, send)
  |   File "/usr/local/lib/python3.11/site-packages/starlette/responses.py", line 258, in __call__
  |     async with anyio.create_task_group() as task_group:
  |   File "/usr/local/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 763, in __aexit__
  |     raise BaseExceptionGroup(
  | ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "/usr/local/lib/python3.11/site-packages/uvicorn/protocols/http/httptools_impl.py", line 401, in run_asgi
    |     result = await app(  # type: ignore[func-returns-value]
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/usr/local/lib/python3.11/site-packages/uvicorn/middleware/proxy_headers.py", line 70, in __call__
    |     return await self.app(scope, receive, send)
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/usr/local/lib/python3.11/site-packages/fastapi/applications.py", line 1054, in __call__
    |     await super().__call__(scope, receive, send)
    |   File "/usr/local/lib/python3.11/site-packages/starlette/applications.py", line 123, in __call__
    |     await self.middleware_stack(scope, receive, send)
    |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/errors.py", line 186, in __call__
    |     raise exc
    |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/errors.py", line 164, in __call__
    |     await self.app(scope, receive, _send)
```

I've also seen `ClientError`, which is strange as it shows up _after_ the import when the function itself is running. I'd expect an `ImportError` if `ClientError` is not found. 

Stranger yet, the 0.1 version I had installed was finding memories but appeared to only save the first one and then would throw an error about `Memory failed`. 

Has anyone gotten this function working? It's the most popular memory one on Open WebUI's site, so I assume so? What settings am I misconfiguring here?

**Comment by joestump** [comment]
Inexplicably, flipping the function to "Global" in Settings yielded some positive results:

```
Added memory: ["User prefers Tesla", "User plans to purchase a Tesla Model S Plaid"]
```

However, the logs continue to show errors:

```
HTTPConnectionPool(host='host.docker.internal', port=11434): Max retries exceeded with url: /api/embed (Caused by NameResolutionError("<urllib3.connection.HTTPConnection object at 0x713757f06ed0>: Failed to resolve 'host.docker.internal' ([Errno -2] Name or service not known)"))

HTTPConnectionPool(host='host.docker.internal', port=11434): Max retries exceeded with url: /api/embed (Caused by NameResolutionError("<urllib3.connection.HTTPConnection object at 0x713758ba2ad0>: Failed to resolve 'host.docker.internal' ([Errno -2] Name or service not known)"))
```

**Comment by joestump** [comment]
Set `OLLAMA_HOST` to `0.0.0.0` and added more logging. The HTTP errors are coming when the function calls [`query_memory`](https://github.com/open-webui/open-webui/blob/c13bcfdfc9fcb9b72e516cb72c89f234511c89df/backend/open_webui/apps/webui/routers/memories.py#L79-L89):

```python
            related_memories = await query_memory(
                request=Request(scope={"type": "http", "app": webui_app}),
                form_data=QueryMemoryForm(
                    content=memory, k=self.valves.related_memories_n
                ),
                user=user,
            )
```

Confused as to why Ollama cannot find its own vector DB. Presumably it makes wide use of this elsewhere?

**Comment by joestump** [comment]
The `ClientError` exceptions are because of a typo in the function as posted to Open-WebUI. You can see in `query_openai_api` it's spelled `ClentError`.

**Comment by joestump** [comment]
Okay, finally figured out the last issue – it turns out that by default the internal Ollama calls in Open-WebUI *ignore* your Ollama configurations in the "Connections" tab of the "Admin Settings". Here's the relevant bits from [the docs](https://docs.openwebui.com/getting-started/advanced-topics/env-configuration#ollama_base_url):

<img width="675" alt="image" src="https://github.com/user-attachments/assets/17a4d5af-93cf-450d-9dcb-6df0fdfa6129">

My Ollama:

* Runs outside of the container
* Runs behind Caddy with TLS enabled
* Is not running on the same host
* Is not running on the default port

I was able to get memories saving correctly by setting `OLLAMA_BASE_URL`. I think this is a bug? Definitely unexpected behavior – if you've configured Ollama via "Connections" in the "Admin Settings", why would this need to be set? Feels like the connection configuration should override the default for `OLLAMA_BASE_URL`?

**Reply by Mavyre** [reply]
If your Ollama runs outside of the container, USE_OLLAMA_DOCKER shoudl't be true.
Also, auto memory uses its own Valve set by default to `http://host.docker.internal:11434`
To make this better, we could replace it by `http://localhost/ollama` and add a Bearer token and it'll use Open Webui Ollama proxy. Or your own Ollama address, if you want it working straightforwardly.
I slightly modified the code to use the same model than the current user conversation. That allows my Ollama to avoid unloading the current model, loading the model for the memory, reunloading, and so on...

**Reply by scottrbaxter** [reply]
This still seems to be an issue (`v0.6.7`) for the embed url (`admin/settings` / `Documents` / `API Base URL under` `Embedding Model Engine` set to `Ollama`) when Ollama is not running on another host (e.g. https://other.host)

I've tried configuring in the UI and also setting variables:
```
ENABLE_PERSISTENT_CONFIG: False
OLLAMA_BASE_URL: https://other.host
RAG_OLLAMA_BASE_URL: https://other.host
RAG_EMBEDDING_ENGINE: ollama
RAG_EMBEDDING_MODEL: nomic-embed-text:137m-v1.5-fp16
USE_OLLAMA_DOCKER: False
```

error:
`urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='other.host', port=443): Max retries exceeded with url: /api/embed (Caused by NameResolutionError("<urllib3.connection.HTTPSConnection object at 0x7048822b9390>: Failed to resolve 'other.host' ([Errno -5] No address associated with hostname)"))`

I've confirmed the open-webui container can curl the ollama host

```
docker exec -it open-webui curl https://other.host
Ollama is running
```
Given the response (and screenshot) above, is there not currently a way to use a custom embedding model with a locally hosted ollama service on any other host?  `http://localhost11434` is the only option? this really feels like a bug.
